src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -187,57 +195,56 @@ object RuleUtils {
   }
 
   /**
-   * Get alternative logical plan of the current plan using the given index.
+   * Transform the current plan to utilize the given index along with newly appended source files.
+   *
    * With HybridScan, indexes with newly appended files to its source relation are also
-   * eligible and we reconstruct new plans for the appended files so as to merge into
-   * the index data.
+   * eligible and we reconstruct new plans for the appended files so as to merge with
+   * bucketed index data correctly.
    *
    * @param spark Spark session.
    * @param index Index used in replacement plan.
    * @param plan Current logical plan.
    * @param useBucketSpec Option whether to use BucketSpec for reading index data.
-   * @return Alternative logical plan.
+   * @return Transformed logical plan.
    */
-  private def transformPlanUsingHybridScan(
+  private def transformPlanToUseHybridIndexDataScan(
       spark: SparkSession,
       index: IndexLogEntry,
       plan: LogicalPlan,
       useBucketSpec: Boolean): LogicalPlan = {
-    // Since the index data is in "parquet" format, we cannot read source files in
-    // formats other than "parquet" using 1 FileScan node (as the operator requires
-    // files in one homogenous format). To address this, we need to read the appended
-    // source files using another FileScan node injected into the plan and subsequently
-    // merge the data into the index data.
-    //
-    // Though BucketUnion (using BucketSpec and on-the-fly Shuffle) is used to merge
-    // them for now, these plans will be optimized with a Union operator at a later time
-    // (see #145).
-    val useBucketUnion = useBucketSpec || !index.relations.head.fileFormat.equals("parquet")
+    val isParquetSourceFormat = index.relations.head.fileFormat.equals("parquet")
     var filesAppended: Seq[Path] = Nil
-    val replacedPlan = plan transformDown {
+
+    // Get replaced plan with index data and appended files if applicable.
+    val indexPlan = plan transformDown {
       case baseRelation @ LogicalRelation(
             _ @HadoopFsRelation(location: PartitioningAwareFileIndex, _, _, _, _, _),
             baseOutput,
             _,
             _) =>
-        // TODO: Duplicate listing files for the given relation as in isHybridScanCandidate.
-        //  See https://github.com/microsoft/hyperspace/issues/160
         val curFileSet = location.allFiles
           .map(f => FileInfo(f.getPath.toString, f.getLen, f.getModificationTime))
-          .toSet
         filesAppended =
-          (curFileSet -- index.allSourceFileInfos).map(f => new Path(f.name)).toSeq
-
+          curFileSet.filterNot(index.allSourceFileInfos.contains).map(f => new Path(f.name))
         // TODO: Hybrid Scan delete support.
```

Comment:
nit: If you are going to immediately address this, it's ok. Otherwise, add an issue here so we can track.